lib/os/mempool.c

```diff
@@ -72,9 +72,16 @@ static size_t buf_size(struct sys_mem_pool_base *p)
 	return p->n_max * p->max_sz;
 }
 
-static bool block_fits(struct sys_mem_pool_base *p, void *block, size_t bsz)
+static bool block_fits(struct sys_mem_pool_base *p,
+		       int lvl, int bn, size_t *lsizes)
 {
-	return ((u8_t *)block + bsz - 1 - (u8_t *)p->buf) < buf_size(p);
+	int plvl = lvl - 1;
+	int pbn = bn / 4;
+	u8_t *block_end = (u8_t *)block_ptr(p, lsizes[lvl], bn) + lsizes[lvl];
+	size_t parent_sz = lvl == 0 ? buf_size(p) : lsizes[plvl];
```

Comment:
suggest parens around lvl == 0 for clarity.
same deal with following line.
or just use a regular if { } else { }. I had to blink at this a lot to correctly parse it.